components/concept/awesomebar/src/main/java/mozilla/components/concept/awesomebar/AwesomeBar.kt

```diff
@@ -0,0 +1,121 @@
+ * Interface to be implemented by awesome bar implementations.
+ *
+ * An awesome bar has multiple duties:
+ *  - Display [Suggestion] instances and invoking its callbacks once selected
+ *  - React to outside events: [onInputStarted], [onInputChanged], [onInputCancelled].
+ *  - Query [SuggestionProvider] instances for new suggestions when the text changes.
+ */
+interface AwesomeBar {
+
+    /**
+     * Adds the following [SuggestionProvider] instances to be queried for [Suggestion]s whenever the text changes.
+     */
+    fun addProviders(vararg providers: SuggestionProvider)
+
+    /**
+     * Fired when the user starts interacting with the awesome bar by entering text in the toolbar.
+     */
+    fun onInputStarted() = Unit
+
+    /**
+     * Fired whenever the user changes their input, after they have started interacting with the awesome bar.
+     *
+     * @param text The current user input in the toolbar.
+     */
+    fun onInputChanged(text: String)
+
+    /**
+     * Fired when the user has cancelled their interaction with the awesome bar.
+     */
+    fun onInputCancelled() = Unit
+
+    /**
+     * Casts this awesome bar to an Android View object.
+     */
+    fun asView(): View = this as View
+
+    /**
+     * Adds a lambda to be invoked when the user has finished interacting with the awesome bar (e.g. selected a
+     * suggestion).
+     */
+    fun setOnStopListener(listener: () -> Unit)
+
+    /**
+     * A [Suggestion] to be displayed by an [AwesomeBar] implementation.
+     *
+     * @property id A unique ID identifying this [Suggestion]. A stable ID but different data indicates to the
+     *              [AwesomeBar] that this is the same [Suggestion] with new data. This will affect how the [AwesomeBar]
+     *              animates showing the new suggestion.
+     * @property title A user-readable title for the [Suggestion].
+     * @property description A user-readable description for the [Suggestion].
+     * @property icon A lambda that can be invoked by the [AwesomeBar] implementation to receive an icon [Bitmap] for
+     *                this [Suggestion]. The [AwesomeBar] will pass in its desired width and height for the Bitmap.
+     * @property chips A list of [Chip] instances to be displayed.
+     * @property flags A set of [Flag] values for this [Suggestion].
+     * @property onSuggestionClicked A callback to be executed when the [Suggestion] was clicked by the user.
+     * @property onChipClicked A callback to be executed when a [Chip] was clicked by the user.
+     */
+    data class Suggestion(
+        val id: String = UUID.randomUUID().toString(),
+        val title: String? = null,
+        val description: String? = null,
+        val icon: ((width: Int, height: Int) -> Bitmap?) = { _, _ -> null },
+        val chips: List<Chip> = emptyList(),
+        val flags: Set<Flag> = emptySet(),
+        val onSuggestionClicked: (() -> Unit)? = null,
+        val onChipClicked: ((Chip) -> Unit)? = null
```

Comment:
More random thoughts: When a chip is clicked, in the future implementation for the Awesome Bar, will that also invoke the AwesomeBar's onStopListener or would that only happen when a suggestion is clicked?

When a chip is clicked, will it need reference to the suggestion it's coming from in the lambda? 🤔

Reply:
Yep! Currently `browser-awesomebar` invokes the listener for both a chip click or a suggestion click. I was wondering whether to return a `Boolean` from the lambda to let the suggestion/chip/provider control that - but so far this wasn't needed.

> When a chip is clicked, will it need reference to the suggestion it's coming from in the lambda? 🤔

Right now that wasn't needed. The `Suggestion` with the `Chip`s are created in the `SuggestionProvider` and there the actual lambda has all the information.

Here are the two providers I implemented so far. One for "open tabs" and one for search suggestions:
https://gist.github.com/pocmo/7f1405c4a8491b09b91e994b23226cdf